### DISCUSSION TITLE
Software hat eine WebUI

### DISCUSSION THREAD

**Discussion by Kenaschon** [dewenni/ESP_Buderus_KM271#22] [General]
Ich habe doch noch eine Frage. Du schreibst die Software hat eine WebUI. Reicht es nur die IP Adresse einzugeben, oder noch einen Port oder so?

Die Software ist aufgespielt und die Einstellungen habe ich angepasst. Der ESP meldet sich auch bei ioBroker mit "esp_heizung". Aber mehr als "info", "status" und "wifi" kommt da nicht.

Grüße
Manfred

**Comment by dewenni** [comment]
Hallo Manfred,

Welche Hardware verwendest du?
Ein Board von Daniel oder einen einzelnen ESP32?
Und welche Heizung?

Weitere Daten kommen natürlich erst wenn auch eine Verbindung zur Heizung besteht.
Am besten kannst du das mit einem MQTT Tool wie dem [MQTT-Explorer](http://mqtt-explorer.com/) beobachten
Das WebUI sollte aber auch so starten. Du kannst über die IP-Adresse ohne weitere Port Angabe die WebUI öffnen.
Wenn da jetzt nichts kommt, wäre das schon etwas komisch.

Zur Sicherheit, du hast in der `config.h` nicht zufällig folgendes auskommentier?
`#define USE_WEBUI `

Ich bin auch kurz davor nochmal eine überarbeitete Version zu veröffentlichen, da ich mit den Ressourcen der WebUI scheinbar an eine Grenze gestoßen bin. Das führt teilweise dazu, dass die WebUI nicht zuverlässig läuft.
Mit der Überarbeitung sollte das dann hoffentlich besser sein. Außerdem kann man dann über die WebUI auch sämtliche Einstellungen vornehmen und muss das nicht mehr im Code machen.

Ich wollte das aber jetzt erstmal noch etwas testen. Auch mit dem neuen Board in der Version 0.0.6
Mal sehen ob ich das am Wochenende schaffe...

Grüße Sven

**Comment by Kenaschon** [comment]
Ich glaube, das ich die vollkommen falsche zip-Datei hatte.

Diese hier:

ESP_Buderus_KM271-aa369b0bc6e71b8ec41ad1284f3467846cb56dcc.zip

einzelnen ESP32

Buderus mit KM271

**Comment by Kenaschon** [comment]
So, jetzt habe ich die aktuelle Software aufgespielt. Komme jetzt auch auf die GUI.

Unter ioBroker bekomme ich jetzt aber folgende Meldungen:

`<html><body>


mqtt.0 | 2023-04-20 17:59:28.857 | warn | State "mqtt.0.esp_heizung.info" has no existing object, this might lead to an error in future versions
-- | -- | -- | --
mqtt.0 | 2023-04-20 17:59:28.805 | warn | State "mqtt.0.esp_heizung.wifi" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:59:18.788 | warn | State "mqtt.0.esp_heizung.info" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:59:18.776 | warn | State "mqtt.0.esp_heizung.wifi" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:59:08.860 | warn | State "mqtt.0.esp_heizung.info" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:59:08.762 | warn | State "mqtt.0.esp_heizung.wifi" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:58:58.839 | warn | State "mqtt.0.esp_heizung.info" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:58:58.739 | warn | State "mqtt.0.esp_heizung.wifi" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:58:48.716 | warn | State "mqtt.0.esp_heizung.info" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:58:48.705 | warn | State "mqtt.0.esp_heizung.wifi" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:58:38.745 | warn | State "mqtt.0.esp_heizung.info" has no existing object, this might lead to an error in future versions
mqtt.0 | 2023-04-20 17:58:38.691 | warn | State "mqtt.0.esp_heizung.wifi" has no existing object, this might lead to an error in future versions


</body>
</html>`

**Comment by dewenni** [comment]
ok, wir kommen der Sache näher :-)

Bitte nutze mal den oben verlinken MQTT Explorer und schau mal wie das dort aussieht.
Ich kenne zwar auch grundsätzlich den ioBroker, bin mir aber unsicher was die Ursache für die Warnungen ist.
Die Topics "wifi" und "info" haben als Inhalt ein JSON Objekt. Vielleicht kann der ioBroker damit nicht umgehen.

sollte dann so aussehen:
<img width="1007" alt="image" src="https://user-images.githubusercontent.com/46074831/233431179-018a625c-5f19-43a9-9364-7c965bd3dfa5.png">

Wenn du das KM271 angeschlossen hast, aber noch keine Daten kommen, dann stimmt mit der Verbindung vielleicht noch etwas nicht.

Mögliche Ursachen:
- GPIO für RX und TX stimmen nicht
- RX / TX vertauscht in der Konfiguration
- gekreuztes Kabel verwendet?

**Comment by Kenaschon** [comment]
Ich habe diesen ESP32

https://community-assets.home-assistant.io/original/3X/7/e/7e769990412b24103a7b87278136ee3ede74c0fa.png

Ist das dann

#define RXD2   3                // ESP32 RX-pin for KM271 communication, align with hardware
#define TXD2   1                // ESP32 TX-pin for KM271 communication, align with hardware

**Comment by dewenni** [comment]
Probiere bitte mal folgende Konfiguration und Verdrahtung:

```
(ESP32)GPIO17/TXD2  -> (Pegelwandler)TXD -> (seriales Kabel) -> (KM271-SUBD)PIN2:RXD
(ESP32)GPIO16/RXD2 <- (Pegelwandler)RXD <- (seriales Kabele) <- (KM271-SUBD)PIN3:TXD
(ESP32)GND <-> (Pegelwandler)GND <-> (seriales Kabel) <-> (KM271-SUBD)PIN5:GND
```

Damit sollte es funktionieren!

**Comment by Kenaschon** [comment]
Nur dadurch hat es aber schon funktioniert.

#define RXD2 3 // ESP32 RX-pin for KM271 communication, align with hardware
#define TXD2 1 // ESP32 TX-pin for KM271 communication, align with hardware

**Comment by Kenaschon** [comment]
Läuft alles einwandfrei. Vom ELV Funksystem FS20 gab es mal das "Wärmebedarfsrelais FHT 8W". Das kann ich jetzt endlich mit der Steuerung wieder nachbauen. Einfach nur die Heizkreispumpe durch ein Relais abschalten ist nie eine gute Idee gewesen. Da löste oft die Sicherheitstemperaturbegrenzung aus.
Jetzt kann ich einfach auf "Nachtbetieb" umschalten, sobald alle Heizkörper aus sind.

Danke für die tolle Umsetzung.

**Comment by Kenaschon** [comment]
Kann es sein, das die Umschaltung, ich benutze HK2, von Automatik auf Nacht per MQTT nicht funktioniert?

Der Datenpunkt ist doch "mqtt.0.esp_heizung.config.HK2_Betriebsart" richtig? Der ändert sich wenn ich in der GUI auf Nachtbetrieb umschalte. Gebe ich dort manuell Nacht ein, ändert sich nichts im Keller an der Heizung.

**Reply by Toastbrot20** [reply]
Ich nutze per MQTT die Umschaltung auf Sommer, wenn alle Heizkörper aus sind.  Das funktioniert gut (Ich nutze ebenfalls den HK2)  Die Pumpe wird dann automatisch abgeschaltet, wenn die Temperatur im Kessel runter gegangen ist.

**Reply by Kenaschon** [reply]
Machst du das auch über ioBroker?

**Reply by Kenaschon** [reply]
Wenn ich das manuell über die GUI mache, funktioniert es, per mqtt nicht.

**Reply by Toastbrot20** [reply]
jepp über iobroker mit nem Blockly script.

**Reply by Toastbrot20** [reply]
ich sehe gerade, dass Dein Datenpunkt falsch ist. Zum senden musst Du neue Datenpunkte anlegen, oder einmal per MQTT was senden, dass die automatisch angelegt werden.

**Reply by Kenaschon** [reply]
Die Datenpunkte werden doch durch den ESP32 in ioBroker automatisch angelegt.

Wenn es hilft. Unter System/Versionsinformationen steht: Software-Version v2.3.0; Logamatic-Version 3.18; Logamatic-Modul 6

**Reply by dewenni** [reply]
@Kenaschon 

schau mal in die param.txt Datei rein. Da sind alle Kommandos aufgelistet.
Für die Betriebsart am HK2 lautet es:

```
command:    heating circuit 2: operation mode 
topic:      {"setvalue/hk2_betriebsart", setvalue/hc2_opmode"}
payload:    (0=night / 1=day / 2=auto)
```

Bei deutscher Sprache also `mqtt.0.esp_heizung.setvalue.hk2_betriebsart`
Und dann den Index 0 füe Nacht, 1 für Tag und 2 für Automatik

Damit der Datenpunkt im ioBroker auftaucht, musst du ihn entweder manuell anlegen oder ggf. mit dem MQTT Explorer mal folgendes publishen `esp_heizung/setvalue/hk2_betriebsart` und als Payload z.B. eine 2

**Reply by dewenni** [reply]
> Die Datenpunkte werden doch durch den ESP32 in ioBroker automatisch angelegt.

Es werden ja nur die angelegt, die von der Heizung gesendet werden. Die Heizung schickt aber nur Status Werte.
Die Config Werte sind im Grunde auch nur Status. Die kann man alle nicht direkt verändern.

Um etwas zu verändern, musst du Nachrichten mit `.../setvalue/...` schicken.
Beispiel: `esp_heizung/setvalue/hk2_betriebsart`

**Reply by Kenaschon** [reply]
Ahh, es werde Licht bzw. Nacht. Jetzt habe ich es verstanden. Vielen Dank für die Hilfe.

Es ist immer wieder toll was manchmal mit solchen doch schon alten Anlagen noch alles möglich ist.

**Reply by Kenaschon** [reply]
> jepp über iobroker mit nem Blockly script.

Fragst du jede Heizung einzeln ab?

**Comment by dewenni** [comment]
Ich habe eben auch noch ein neues release veröffentlicht. Schaut mal rein und gebt gerne Rückmeldung ob es bei euch läuft.

**Comment by Kenaschon** [comment]
Ist das richtig, das die Datei "Credentials.h" fehlt?

**Reply by dewenni** [reply]
ja, das habe ich alles umgebaut.
Sämtliche Einstellungen werden jetzt über das webUI vorgenommen.
Wenn du die neue Software erstmals aufspielt, wird der ESP in den neuen "Setup-Mode" gehen, weil er keine Config Datei findet.
Es wird dann ein Accesspoint mit dem Namen "ESP-Buderus-KM271" aufgemacht.
Mit dem musst du dich dann verbinden und danach auf die Seite http://192.168.4.1 gehen.
Dort kannst du dann alles eintragen und danach den Button "Save and Restart" drücken.

Bei zukünftigen Updates, bleiben die Einstellungen dann erhalten. Diese werden als Datei im Dateisystem gespeichert.

**Reply by Kenaschon** [reply]
So, habe es aufgespielt. Leider bekomme ich keine Werte von der Heizung.

**Reply by dewenni** [reply]
@Kenaschon 
hast du denn in den Einstellungen die GPIO´s eingestellt?

<img width="506" alt="image" src="https://user-images.githubusercontent.com/46074831/233780372-60abe57e-2ef7-48b3-a324-88730f5c5a05.png">

du hattest zuvor:
> #define RXD2 3 // ESP32 RX-pin for KM271 communication, align with hardware
#define TXD2 1 // ESP32 TX-pin for KM271 communication, align with hardware

also musst du dort folgendes einstellen: 
```
KM271-RX = 3
KM271-TX = 1 
```

den Rest kannst du auf -1 stellen wenn du keine LED´s oder Ölzähler angeschlossen hast.

**Comment by Kenaschon** [comment]
An welche PINS muß ich den drann? Mit der alten Version war ich an TXD - GPI01 und RXD - GPI03 (oben rechts im Bild, innere Pinreihe)

![esp32mini](https://user-images.githubusercontent.com/47242804/233776199-bcf3cbf5-a6c2-450a-b5df-cd1e93ba5b8a.jpg)

**Reply by Kenaschon** [reply]
Alte Version aufgespielt - Werte alle da.

**Reply by Kenaschon** [reply]
Warum auch immer - ich habe es jetzt noch einmal aufgespielt - jetzt ist alles da.

**Comment by Kenaschon** [comment]
Kann es sein, das der Ölzähler noch nicht funktioniert?

Habe in den Einstellungen auf "virtuell" stehen. Verbrauch und Öldichte eingestellt. In den Daten die per mqtt kommen steht auch ein Oelverbrauch_Gesamt_berechnet.

In der WEBUI auf dem Dashboard bei Ölzähler sind immer zwei Striche. Steht dort nur etwas wenn ein richtiger Ölzähler angeschlossen ist?

**Reply by dewenni** [reply]
Danke, da hatte ich tatsächlich noch einen Fehler.
Habe es gefixed und hochgeladen. Mache dafür aber noch kein neues Release.
Du kannst die aber den neusten Stand auch so nochmal runterladen.

**Comment by Kenaschon** [comment]
Noch eine Frage. Gibt es eine Einstellung, bei der das Warmwasser auf Temperatur gehalten wird, HK2-Betriebsart auf Nacht steht?

**Reply by dewenni** [reply]
Du kannst mal testen was passiert, wenn du die Betriebsart vom Warmwasser auf "Tag" stellst, anstelle von Automatik.
Dann wird es vermutlich auf dann auf Temperatur gehalten, wenn der Heizkreis auf Nacht geht.
Habe ich bisher aber noch nicht getestet.

**Comment by Kenaschon** [comment]
Noch eine Frage.

Ist das ein Fehler bei der Werteübergabe durch mqtt oder was bedeuten die Zahlen?

![Unbenannt-1](https://user-images.githubusercontent.com/47242804/233939067-3d8a4128-ce23-4677-b6b9-257049b6d192.jpg)

Ich muß schon sagen - alles eine wirklich tolle Arbeit. Hut ab. Danke dafür.

**Reply by dewenni** [reply]
nein, das stimmt schon so.
Die Laufzeit wird in der Logamatic auf mehrere Datenpunkte aufgeteilt.
Brenner_Laufzeit_Minuten = enthält die aktuellen Minuten
Brenner_Laufzeit_Minuten256 = wird alle 256 Minuten um 1 erhöht
Brenner_Laufzeit_Minuten65536 = wird alle 65536 Minuten um 1 erhöht

Die Gesamtlaufzeit ist dann die Summe aus:   
(Brenner_Laufzeit_Minuten65536 * 65536) + (Brenner_Laufzeit_Minuten256 * 256) * Brenner_Laufzeit_Minuten

Das rechne ich im ESP schon zusammen und gebe es als Brenner_Laufzeit_Summe aus

**Comment by Kenaschon** [comment]
Noch eine Frage. Der Wert für die Abgastemperatur steht immer auf  255. 

Kommt da kein aktueller Wert? Muß dort ein Sensor eingebaut werden?

**Reply by dewenni** [reply]
ja, da wird normalerweise noch ein Abgasfühler angeschlossen.
Das Board von Daniel hat dafür auch einen Stecker vorgesehen.
Das Toggelt bei mir auch zwischen 0 und 255

**Comment by Kenaschon** [comment]
Und noch eine Frage.

Kann es sein, das die Werte von HK2_Vorlaufisttemperatur und HK2_Vorlaufsolltemperatur mit denen von Kessel_Vorlaufisttemperatur und Kessel_Vorlaufsolltemperatur vertauscht sind?

![vorlauf](https://user-images.githubusercontent.com/47242804/233950184-6fd77e8e-9ae2-4917-8a86-5bce933f05b1.jpg)

Bei mir an der Heizung zeigt das Termometer im Vorlauf zu den Heizungen immer so um 70 C° an.

Ist das nicht der Wert für HK2_Vorlaufisttemperatur?

**Reply by dewenni** [reply]
von den Werten klingt das für mich logisch.
Die Kesseltemperaturen sind ja in der Regel höher als die Vorlauftemperaturen.

Während der Brenner eingeschaltet ist, sind die Werte glaube ich sogar gleich.
Nachdem der Brenner dann ausgeschaltet hat, sind die wieder unterschiedlich.

**Comment by Kenaschon** [comment]
Ok.

**Comment by Kenaschon** [comment]
Ich habe da noch eine Frage zu deiner Grafana-Grafik.
Kannst du kurz erklären, nur wenn es geht, wie du den Zustand von zwei Datenpunkten in eine Grafik bekommen hast.

![Unbenannt-2](https://user-images.githubusercontent.com/47242804/234038950-0727dfcc-3918-4197-8878-1c4657e0c044.jpg)

Wie hier bei Warmwasser. Das ergibt sich ja aus zwei Datenpunkte. Automatik( Auto=1; Tag=0) - Tag(Auto=0; Tag=1) - Nacht(Auto=0; Tag=0).

Wertest du das mit einem kleinen Skript dazwischen aus?

**Reply by dewenni** [reply]
Ich nutze dort die Datenpunkte aus den Betriebswerten1

Für die Anzeige beim Heizkreis 1 z.B.  `HK1_BW1_Automatik`

Für die Anzeige beim Warmwasser  `WW_BW1_Automatik`

Es wird also nur Angezeigt ob Automatik aktiv ist. Wenn nicht, dann wird Handbetrieb angezeigt.
In der WebUI habe ich das etwas anders gemacht.
Da wird bei Handbetrieb auch zusätzlich noch Tag oder Nacht angezeigt.

**Comment by Kenaschon** [comment]
Noch mal eine Frage.

Mit `mqtt.0.Heizung/setvalue/hk2_betriebsart` kann ich mit mqtt von Automatik auf Nacht und Tag umschalten.

Für WW-Betriebsart geht das nicht. Wenn ich hiermit `mqtt.0.Heizung/setvalue/ww_betriebsart` versuche die WW-Betriebsart zu ändern passiert nichts.

Wo habe ich den Fehler?

**Reply by dewenni** [reply]
Das muss ich mir ansehen. Schaffe ich aber frühestens am Freitag.

Das Topic sollte aber passen.
Funktioniert es denn über das WebUI?

**Reply by Kenaschon** [reply]
Jepp, dort geht es.

**Reply by dewenni** [reply]
Und als Payload übergibst du 0,1,2 ?

**Reply by Kenaschon** [reply]
Wie es hier steht

`command:    warm water: operation mode
topic:      {"setvalue/ww_betriebsart", setvalue/ww_opmode"}
payload:    0=night / 1=day / 2=auto`

**Reply by dewenni** [reply]
Ich habe es gerade getestet und es sieht zumindest bei mir so aus, als würde es funktionieren.

_**Umschaltung auf manuell Nacht:**_
Topic: .../setvalue/ww_betriebsart
Payload: 0

_**Umschaltung auf manuell Tag:**_  
Topic: .../setvalue/ww_betriebsart  
Payload: 1  

_**Umschaltung auf Automatik:**_
Topic: .../setvalue/ww_betriebsart
Payload: 2

**Reply by dewenni** [reply]
lediglich bei der message die ich als Information zurück sende, wenn ein Kommando empfangen wurde, war noch ein copy/paste Fehler drin. Da kam der Text: "hk2_programm - empfangen" anstelle von "ww_opmode - empfangen"
Das hatte aber keine Auswirkung auf die Funktion.

**Reply by Kenaschon** [reply]
Warum auch immer. Jetzt geht es. Anscheinend ist der "Unterstrich" beim Datenpunkt erzeugen nicht richtig mitgekommen.

Jetzt löpp es.

Danke für die Hilfe.

**Comment by dewenni** [comment]
Ich schließe die Diskussion mal an diese Stelle.
Der ursprüngliche Titel ist ja schon längst nicht mehr Thema.
Wenn es weitere Fragen gib, gerne nochmal eine Diskussion eröffnen.
